packages/editor/src/lib/exports/FontEmbedder.ts

```diff
@@ -0,0 +1,101 @@
+import { assert, bind, compact } from '@tldraw/utils'
+import { Promise } from 'core-js'
+import { fetchCache, resourceToDataUrl } from './fetchCache'
+import { ParsedFontFace, parseCss, parseCssFontFaces, parseCssFontFamilyValue } from './parseCss'
+
+export class FontEmbedder {
+	private fontFacesPromise: Promise<ParsedFontFace[]> | null = null
+	private readonly foundFontNames = new Set<string>()
+	private readonly fontFacesToEmbed = new Set<ParsedFontFace>()
+	private readonly pendingPromises: Promise<void>[] = []
+
+	startFindingCurrentDocumentFontFaces() {
+		assert(!this.fontFacesPromise, 'FontEmbedder already started')
+		this.fontFacesPromise = getCurrentDocumentFontFaces()
+	}
+
+	@bind onFontFamilyValue(fontFamilyValue: string) {
+		assert(this.fontFacesPromise, 'FontEmbedder not started')
+
+		const fonts = parseCssFontFamilyValue(fontFamilyValue)
+		for (const font of fonts) {
+			if (this.foundFontNames.has(font)) return
+			this.foundFontNames.add(font)
+
+			this.pendingPromises.push(
+				this.fontFacesPromise.then((fontFaces) => {
+					const relevantFontFaces = fontFaces.filter((fontFace) => fontFace.fontFamilies.has(font))
+					for (const fontFace of relevantFontFaces) {
+						if (this.fontFacesToEmbed.has(fontFace)) continue
+
+						this.fontFacesToEmbed.add(fontFace)
+						for (const url of fontFace.urls) {
+							if (!url.resolved || url.embedded) continue
+							// kick off fetching this font
+							url.embedded = resourceToDataUrl(url.resolved)
+						}
+					}
+				})
+			)
+		}
+	}
+
+	async createCss() {
+		await Promise.all(this.pendingPromises)
+
+		let css = ''
+
+		for (const fontFace of this.fontFacesToEmbed) {
+			let fontFaceString = `@font-face {${fontFace.fontFace}}`
+
+			for (const url of fontFace.urls) {
+				if (!url.embedded) continue
+				const dataUrl = await url.embedded
+				if (!dataUrl) continue
+
+				fontFaceString = fontFaceString.replace(url.original, dataUrl)
```

Comment:
worth mentioning with a comment here what this accomplishing, or perhaps better just an overview comment at the top of the file that you're doing some embedding here

Reply:
added a comment at the top giving an overview of what this class is for and how it works